storage/replica_command.go

```diff
@@ -1255,45 +1255,71 @@ func (r *Replica) splitTrigger(batch engine.Engine, split *proto.SplitTrigger) e
 func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDescriptor) (proto.AdminMergeResponse, error) {
 	var reply proto.AdminMergeResponse
 
-	// Lookup subsumed range.
 	if desc.EndKey.Equal(proto.KeyMax) {
-		// Noop.
-		return reply, nil
-	}
-	subsumedRng := r.rm.LookupReplica(desc.EndKey, nil)
-	if subsumedRng == nil {
-		return reply, util.Errorf("ranges not collocated; migration of ranges in anticipation of merge not yet implemented")
+		// Merging the final range doesn't make sense.
+		return reply, util.Errorf("cannot merge final range")
 	}
-	subsumedDesc := subsumedRng.Desc()
 
-	// Make sure the range being subsumed follows this one.
-	if !bytes.Equal(desc.EndKey, subsumedDesc.StartKey) {
-		return reply, util.Errorf("Ranges that are not adjacent cannot be merged, %s != %s", desc.EndKey, subsumedDesc.StartKey)
-	}
+	// Lookup subsumed range. This really belongs inside the transaction
+	// for consistency, but it is important (for transaction record placement)
+	// that the first action inside the transaction is the conditional put
+	// to change the first descriptor's end key. We look up the descriptor
+	// here only to get the new end key and then repeat the lookup inside the
+	// transaction.
+	var newEndKey proto.Key
+	{
+		subsumedRng := r.rm.LookupReplica(desc.EndKey, nil)
+		if subsumedRng == nil {
+			return reply, util.Errorf("ranges not collocated")
```

Comment:
can this error message provide a bit more context? "range containing key %s not collocated with with range containing %s" or whatever other information might be friendly